APMrover2/sailboat.cpp

```diff
@@ -87,48 +96,124 @@ Sailboat::Sailboat()
+}
+
 void Sailboat::init()
 {
     // sailboat defaults
-    if (enabled()) {
+    if (sail_enabled()) {
         rover.g2.crash_angle.set_default(0);
+    }
+
+    if (tack_enabled()) {
         rover.g2.loit_type.set_default(1);
         rover.g2.loit_radius.set_default(5);
         rover.g2.wp_nav.set_default_overshoot(10);
     }
+
+    // initialise motor state to USE_MOTOR_ASSIST
+    // this will silently fail if there is no motor attached
+    set_motor_state(UseMotor::USE_MOTOR_ASSIST, false);
+}
+
+// initialise rc input (channel_mainsail), may be called intermittently
+void Sailboat::init_rc_in()
+{
+    // get auxiliary throttle value
+    RC_Channel *rc_ptr = rc().find_channel_for_option(RC_Channel::AUX_FUNC::MAINSAIL);
+    if (rc_ptr != nullptr) {
+        // use aux as sail input if defined
+        channel_mainsail = rc_ptr;
+        channel_mainsail->set_angle(100);
+        channel_mainsail->set_default_dead_zone(30);
+    } else {
+        // use throttle channel
+        channel_mainsail = rover.channel_throttle;
+    }
 }
 
-// update mainsail's desired angle based on wind speed and direction and desired speed (in m/s)
-void Sailboat::update_mainsail(float desired_speed)
+// decode pilot mainsail input and return in steer_out and throttle_out arguments
+// mainsail_out is in the range 0 to 100, defaults to 100 (fully relaxed) if no input configured
+void Sailboat::get_pilot_desired_mainsail(float &mainsail_out)
 {
-    if (!enabled()) {
+    // no RC input means mainsail is moved to trim
+    if ((rover.failsafe.bits & FAILSAFE_EVENT_THROTTLE) || (channel_mainsail == nullptr)) {
+        mainsail_out = 100.0f;
         return;
     }
+    mainsail_out = constrain_float(channel_mainsail->get_control_in(), 0.0f, 100.0f);
+}
 
-    // relax sail if desired speed is zero
-    if (!is_positive(desired_speed)) {
-        rover.g2.motors.set_mainsail(100.0f);
+// calculate throttle and mainsail angle required to attain desired speed (in m/s)
+// returns true if successful, false if sailboats not enabled
+void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttle_out, float &mainsail_out)
+{
+    if (!sail_enabled()) {
+        throttle_out = 0.0f;
+        mainsail_out = 0.0f;
         return;
     }
 
-    // + is wind over starboard side, - is wind over port side, but as the sails are sheeted the same on each side it makes no difference so take abs
-    float wind_dir_apparent = fabsf(rover.g2.windvane.get_apparent_wind_direction_rad());
-    wind_dir_apparent = degrees(wind_dir_apparent);
+    // run speed controller if motor is forced on or motor assistance is required for low speeds or tacking
+    if ((motor_state == UseMotor::USE_MOTOR_ALWAYS) ||
+         motor_assist_tack() ||
+         motor_assist_low_wind()) {
+        // run speed controller - duplicate of calls found in mode::calc_throttle();
+        throttle_out = 100.0f * rover.g2.attitude_control.get_throttle_out_speed(desired_speed,
```

Comment:
we the ability to call get_throttle_out_stop for sailboats even when forcing motoring

Reply:
Indeed, but get_throttle_out_stop isn't really meant for boats anyway because it leads to the vehicle drifting away.  It works for Rovers because shutting down the motors normally makes them stop moving all all... but we don't use it for boats as far as I know.